File: lib/djv/App/FileActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the DJV project.

#include <djv/App/FileActions.h>

#include <djv/App/App.h>
#include <djv/Models/FilesModel.h>

#include <ftk/Core/Format.h>
#include <ftk/Core/String.h>


#include <algorithm>

namespace djv
{
    namespace app
    {
        namespace
        {
            // The file a command acts on: the "A" file, or the one at the
            // index given as "file".
            std::shared_ptr<models::FilesModelItem> commandFile(
                const std::shared_ptr<App>& app,
                const nlohmann::json& args)
            {
                const auto& files = app->getFilesModel()->getFiles();
                if (args.is_object() && args.contains("file"))
                {
                    const int index = args.at("file").get<int>();
                    if (index < 0 || index >= static_cast<int>(files.size()))
                    {
                        throw std::invalid_argument(ftk::Format(
                            "No file {0}; there are {1}").arg(index).arg(files.size()));
                    }
                    return files[index];
                }
                auto a = app->getFilesModel()->getA();
                if (!a)
                {
                    throw std::invalid_argument("No file is open");
                }
                return a;
            }
        }

        struct FileActions::Private
        {
            std::shared_ptr<ftk::ListObserver<std::shared_ptr<models::FilesModelItem> > > filesObserver;
            std::shared_ptr<ftk::Observer<std::shared_ptr<models::FilesModelItem> > > aObserver;
        };

        void FileActions::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            IActions::_init(context, app, "File");
            FTK_P();

            // Register the commands.
            auto appWeak = std::weak_ptr<App>(app);
            _addCommand(
                "Open",
                "Open a file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->openDialog();
                    }
                });

            _addCommand(
                "OpenAudio",
                "Open a file with a separate audio file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->openSeparateAudioDialog();
                    }
                });

            _addCommand(
                "OpenPlaylist",
                "Open a playlist into the file list. Takes an optional "
                "\"fileName\"; without one a file browser is shown.",
                [appWeak](const nlohmann::json& args)
                {
                    if (auto app = appWeak.lock())
                    {
                        if (args.contains("fileName"))
                        {
                            app->openPlaylist(ftk::Path(
                                args.at("fileName").get<std::string>()));
                        }
                        else
                        {
                            app->openPlaylistDialog();
                        }
                    }
                });

            _addCommand(
                "SavePlaylist",
                "Save the file list as a playlist. Takes an optional "
                "\"fileName\"; without one a file browser is shown.",
                [appWeak](const nlohmann::json& args)
                {
                    if (auto app = appWeak.lock())
                    {
                        if (args.contains("fileName"))
                        {
                            app->savePlaylist(ftk::Path(
                                args.at("fileName").get<std::string>()));
                        }
                        else
                        {
                            app->savePlaylistDialog();
                        }
                    }
                });

            _addCommand(
                "Close",
                "Close the current file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->closeFile();
                    }
                });

            _addCommand(
                "CloseAll",
                "Close all files.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->closeAllFiles();
                    }
                });

            _addCommand(
                "Reload",
                "Reload the current file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->reload();
                    }
                });

            _addCommand(
                "Next",
                "Change to the next file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->getFilesModel()->next();
                    }
                });

            _addCommand(
                "Prev",
                "Change to the previous file.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->getFilesModel()->prev();
                    }
                });

            _addCommand(
                "NextMediaReference",
                "Change to the next media reference.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->getFilesModel()->nextMediaReferenceKey();
                    }
                });

            // By name, for scripts: the cycle above depends on which keys a
            // timeline happens to use.
            _addCommand(
                "MediaReference",
                "Set a file's media reference by name, e.g., { \"key\": \"Proxy\" }; "
                "\"As Authored\" or an empty key is the references the clips were "
                "authored with. It is the \"A\" file's unless \"file\" gives the "
                "index of another in the file list.",
                [appWeak](const nlohmann::json& args)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto item = commandFile(app, args);
                        const std::string key = args.is_object() && args.contains("key") ?
                            args.at("key").get<std::string>() :
                            std::string();
                        // A file that has not been opened yet takes the key
                        // as it is, and checks it when it opens.
                        if (!item->mediaReferenceKeysKnown)
                        {
                            app->getFilesModel()->setMediaReferenceKey(item, key);
                            return;
                        }
                        if (item->mediaReferenceKeys.empty() && !key.empty())
                        {
                            throw std::invalid_argument(ftk::Format(
                                "{0} has only the one media reference").
                                arg(item->path.getFileName()));
                        }
                        const auto found = models::findMediaReferenceKey(*item, key);
                        if (!found.has_value())
                        {
                            std::vector<std::string> names;
                            for (const auto& i : item->mediaReferenceKeys)
                            {
                                names.push_back(models::getMediaReferenceLabel(i));
                            }
                            throw std::invalid_argument(ftk::Format(
                                "{0} has no media reference \"{1}\"; it has {2}").
                                arg(item->path.getFileName()).
                                arg(key).
                                arg(ftk::join(names, ", ")));
                        }
                        app->getFilesModel()->setMediaReferenceKey(item, found.value());
                    }
                });

            _addCommand(
                "Layer",
                "Set a file's layer by name or index, e.g., { \"layer\": \"diffuse\" }. "
                "It is the \"A\" file's unless \"file\" gives the index of another "
                "in the file list.",
                [appWeak](const nlohmann::json& args)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto item = commandFile(app, args);
                        if (!args.is_object() || !args.contains("layer"))
                        {
                            throw std::invalid_argument("No \"layer\" given");
                        }
                        const auto& value = args.at("layer");
                        const std::string layer = value.is_number_integer() ?
                            std::to_string(value.get<int64_t>()) :
                            value.get<std::string>();
                        const auto found = models::findLayer(*item, layer);
                        if (!found.has_value())
                        {
                            throw std::invalid_argument(ftk::Format(
                                "{0} has no layer \"{1}\"; it has {2}").
                                arg(item->path.getFileName()).
                                arg(layer).
                                arg(ftk::join(item->videoLayers, ", ")));
                        }
                        app->getFilesModel()->setLayer(item, static_cast<int>(found.value()));
                    }
                });

            _addCommand(
                "NextLayer",
                "Change to the next layer.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->getFilesModel()->nextLayer();
                    }
                });

            _addCommand(
                "PrevLayer",
                "Change to the previous layer.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        app->getFilesModel()->prevLayer();
                    }
                });

            _addCommand(
                "Exit",
                "Exit the application.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        // exit() prompts about an unsaved review itself, for
                        // every way of quitting.
                        app->exit();
                    }
                });

            // Create the actions.
            _actions["Open"] = ftk::Action::create(
                "Open",
                "FileOpen",
                _command("Open"));
            _actions["OpenAudio"] = ftk::Action::create(
                "Open With Audio",
                "FileOpenAudio",
                _command("OpenAudio"));
            _actions["OpenPlaylist"] = ftk::Action::create(
                "Open Playlist",
                _command("OpenPlaylist"));
            // In the tooltips map rather than set on the action, so the
            // shortcuts update does not put the command's plainer wording
            // back over it.
            _tooltips["OpenPlaylist"] =
                "Open a playlist into the file list. Opening a \".otio\" "
                "file normally plays it as a timeline.";
            _actions["SavePlaylist"] = ftk::Action::create(
                "Save Playlist",
                _command("SavePlaylist"));
            _tooltips["SavePlaylist"] =
                "Save the file list as a \".otio\" playlist.";
            _actions["Close"] = ftk::Action::create(
                "Close",
                "FileClose",
                _command("Close"));
            _actions["CloseAll"] = ftk::Action::create(
                "Close All",
                "FileCloseAll",
                _command("CloseAll"));
            _actions["Reload"] = ftk::Action::create(
                "Reload",
                "FileReload",
                _command("Reload"));
            _actions["Next"] = ftk::Action::create(
                "Next",
                "Next",
                _command("Next"));
            _actions["Prev"] = ftk::Action::create(
                "Previous",
                "Prev",
                _command("Prev"));
            _actions["NextMediaReference"] = ftk::Action::create(
                "Next Media Reference",
                "Next",
                _command("NextMediaReference"));
            _actions["NextLayer"] = ftk::Action::create(
                "Next Layer",
                "Next",
                _command("NextLayer"));
            _actions["PrevLayer"] = ftk::Action::create(
                "Previous Layer",
                "Prev",
                _command("PrevLayer"));
            _actions["Exit"] = ftk::Action::create(
                "Exit",
                _command("Exit"));

            // Register the shortcuts.
            _addShortcut("Open", ftk::KeyShortcut(ftk::Key::O, static_cast<int>(ftk::commandKeyModifier)));
            _addShortcut("OpenAudio",
                ftk::KeyShortcut(
                    ftk::Key::O,
                    static_cast<int>(ftk::KeyModifier::Shift) |
                    static_cast<int>(ftk::commandKeyModifier)));
            _addShortcut("OpenPlaylist", "Open playlist");
            _addShortcut("SavePlaylist", "Save playlist");
            _addShortcut("Close", ftk::KeyShortcut(ftk::Key::E, static_cast<int>(ftk::commandKeyModifier)));
            _addShortcut("CloseAll",
                ftk::KeyShortcut(
                    ftk::Key::E,
                    static_cast<int>(ftk::KeyModifier::Shift) | static_cast<int>(ftk::commandKeyModifier)));
            _addShortcut("Reload",
                ftk::KeyShortcut(
                    ftk::Key::R,
                    static_cast<int>(ftk::KeyModifier::Shift) | static_cast<int>(ftk::commandKeyModifier)));
            _addShortcut("Next", ftk::KeyShortcut(ftk::Key::PageDown, static_cast<int>(ftk::KeyModifier::Control)));
            _addShortcut("Prev", ftk::KeyShortcut(ftk::Key::PageUp, static_cast<int>(ftk::KeyModifier::Control)));
            _addShortcut("NextMediaReference",
                ftk::KeyShortcut(ftk::Key::M, static_cast<int>(ftk::KeyModifier::Shift)));
            _addShortcut("NextLayer", ftk::KeyShortcut(ftk::Key::Equals, static_cast<int>(ftk::KeyModifier::Control)));
            _addShortcut("PrevLayer", ftk::KeyShortcut(ftk::Key::Minus, static_cast<int>(ftk::KeyModifier::Control)));
            _addShortcut("Exit", ftk::KeyShortcut(ftk::Key::Q, static_cast<int>(ftk::commandKeyModifier)));

            _shortcutsUpdate(app->getSettingsModel()->getShortcuts());

            p.filesObserver = ftk::ListObserver<std::shared_ptr<models::FilesModelItem> >::create(
                app->getFilesModel()->observeFiles(),
                [this](const std::vector<std::shared_ptr<models::FilesModelItem> >& value)
                {
                    FTK_P();
                    _actions["SavePlaylist"]->setEnabled(!value.empty());
                    _actions["Close"]->setEnabled(!value.empty());
                    _actions["CloseAll"]->setEnabled(!value.empty());
                    _actions["Reload"]->setEnabled(!value.empty());
                    _actions["Next"]->setEnabled(value.size() > 1);
                    _actions["Prev"]->setEnabled(value.size() > 1);
                });

            p.aObserver = ftk::Observer<std::shared_ptr<models::FilesModelItem> >::create(
                app->getFilesModel()->observeA(),
                [this](const std::shared_ptr<models::FilesModelItem>& value)
                {
                    _actions["NextLayer"]->setEnabled(value ? value->videoLayers.size() > 1 : false);
                    _actions["PrevLayer"]->setEnabled(value ? value->videoLayers.size() > 1 : false);
                    // The keys are only listed when there is a choice.
                    _actions["NextMediaReference"]->setEnabled(
                        value ? !value->mediaReferenceKeys.empty() : false);
                });

        }

        FileActions::FileActions() :
            _p(new Private)
        {}

        FileActions::~FileActions()
        {}

        std::shared_ptr<FileActions> FileActions::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            auto out = std::shared_ptr<FileActions>(new FileActions);
            out->_init(context, app);
            return out;
        }
    }
}

File: examples/python/FileActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the DJV project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl
import djvPy as djv

import IActions
import Util

import weakref

class Actions(IActions.IActions):
    """
    This class provides file actions.
    """
    def __init__(self, context, app, mainWindow):
        IActions.IActions.__init__(self, context, app, "File")

        appWeak = weakref.ref(app)
        mainWindowWeak = weakref.ref(mainWindow)

        # Register the commands.
        self._addCommand(
            "Open",
            "Open a file.",
            lambda args: context.getSystemByName("ftk::FileBrowserSystem").open(
                mainWindowWeak(),
                appWeak().open))

        self._addCommand(
            "OpenAudio",
            "Open a file with a separate audio file.",
            lambda args: appWeak().openSeparateAudioDialog())

        self._addCommand(
            "OpenPlaylist",
            "Open a playlist into the file list.",
            lambda args: appWeak().openPlaylist(ftk.Path(args["fileName"]))
                if args and "fileName" in args
                else appWeak().openPlaylistDialog())

        self._addCommand(
            "SavePlaylist",
            "Save the file list as a playlist.",
            lambda args: appWeak().savePlaylist(ftk.Path(args["fileName"]))
                if args and "fileName" in args
                else appWeak().savePlaylistDialog())

        self._addCommand(
            "Close",
            "Close the current file.",
            lambda args: appWeak().getFilesModel().close())

        self._addCommand(
            "CloseAll",
            "Close all files.",
            lambda args: appWeak().getFilesModel().closeAll())

        self._addCommand(
            "Reload",
            "Reload the current file.",
            lambda args: appWeak().reload())

        self._addCommand(
            "Next",
            "Change to the next file.",
            lambda args: appWeak().getFilesModel().next())

        self._addCommand(
            "Prev",
            "Change to the previous file.",
            lambda args: appWeak().getFilesModel().prev())

        self._addCommand(
            "NextMediaReference",
            "Change to the next media reference.",
            lambda args, f = Util.weak(self._nextMediaReference): f())

        self._addCommand(
            "NextLayer",
            "Change to the next layer.",
            lambda args: appWeak().getFilesModel().nextLayer())

        self._addCommand(
            "PrevLayer",
            "Change to the previous layer.",
            lambda args: appWeak().getFilesModel().prevLayer())

        self._addCommand(
            "OpenReview",
            "Open a review, replacing the current session.",
            lambda args: appWeak().openReview(args["fileName"])
                if args and "fileName" in args
                else appWeak().openReviewDialog())

        self._addCommand(
            "SaveReview",
            "Save the current session as a review.",
            lambda args: appWeak().saveReview())

        self._addCommand(
            "SaveReviewAs",
            "Save the current session as a new review.",
            lambda args: appWeak().saveReviewAs())

        self._addCommand(
            "CloseReview",
            "Close the review and reset to the startup state.",
            lambda args: appWeak().closeReview())

        self._addCommand(
            "Exit",
            "Exit the application.",
            lambda args: appWeak().exit())

        # Create the actions.
        self.actions["Open"] = ftk.Action(
            "Open",
            "FileOpen",
            self._command("Open"))
        self.actions["OpenAudio"] = ftk.Action(
            "Open With Audio",
            "FileOpenAudio",
            self._command("OpenAudio"))
        self.actions["OpenPlaylist"] = ftk.Action(
            "Open Playlist",
            self._command("OpenPlaylist"))
        self.actions["OpenPlaylist"].tooltip = (
            "Open a playlist into the file list. Opening a \".otio\" "
            "file normally plays it as a timeline.")
        self.actions["SavePlaylist"] = ftk.Action(
            "Save Playlist",
            self._command("SavePlaylist"))
        self.actions["SavePlaylist"].tooltip = (
            "Save the file list as a \".otio\" playlist.")
        self.actions["OpenReview"] = ftk.Action(
            "Open Review",
            self._command("OpenReview"))
        self.actions["SaveReview"] = ftk.Action(
            "Save Review",
            self._command("SaveReview"))
        self.actions["SaveReviewAs"] = ftk.Action(
            "Save Review As...",
            self._command("SaveReviewAs"))
        self.actions["CloseReview"] = ftk.Action(
            "Close Review",
            self._command("CloseReview"))
        self.actions["Close"] = ftk.Action(
            "Close",
            "FileClose",
            self._command("Close"))
        self.actions["CloseAll"] = ftk.Action(
            "Close All",
            "FileCloseAll",
            self._command("CloseAll"))
        self.actions["Reload"] = ftk.Action(
            "Reload",
            "FileReload",
            self._command("Reload"))
        self.actions["Next"] = ftk.Action(
            "Next",
            "Next",
            self._command("Next"))
        self.actions["Prev"] = ftk.Action(
            "Previous",
            "Prev",
            self._command("Prev"))
        self.actions["NextMediaReference"] = ftk.Action(
            "Next Media Reference",
            "Next",
            self._command("NextMediaReference"))
        self.actions["NextLayer"] = ftk.Action(
            "Next Layer",
            "Next",
            self._command("NextLayer"))
        self.actions["PrevLayer"] = ftk.Action(
            "Previous Layer",
            "Prev",
            self._command("PrevLayer"))
        self.actions["Exit"] = ftk.Action(
            "Exit",
            self._command("Exit"))

        # Register the shortcuts.
        self._addShortcut("Open", ftk.KeyShortcut(ftk.Key.O, ftk.commandKeyModifier))
        self._addShortcut("OpenAudio", ftk.KeyShortcut(
            ftk.Key.O, ftk.KeyModifier.Shift, ftk.commandKeyModifier))
        # Alt rather than Shift on the command modifier: Shift+Ctrl+O is
        # "Open with audio", and Shift+Ctrl+S is free but keeping the
        # pair symmetrical is worth more than reusing it.
        self._addShortcut("OpenReview", "Open review", ftk.KeyShortcut(
            ftk.Key.O, ftk.KeyModifier.Alt, ftk.commandKeyModifier))
        self._addShortcut("SaveReview", "Save review", ftk.KeyShortcut(
            ftk.Key.S, ftk.KeyModifier.Alt, ftk.commandKeyModifier))
        self._addShortcut("SaveReviewAs", "Save review as")
        self._addShortcut("CloseReview", "Close review")
        self._addShortcut("Close", ftk.KeyShortcut(ftk.Key.E, ftk.commandKeyModifier))
        self._addShortcut("CloseAll", ftk.KeyShortcut(
            ftk.Key.E, ftk.KeyModifier.Shift, ftk.commandKeyModifier))
        self._addShortcut("Reload", ftk.KeyShortcut(
            ftk.Key.R, ftk.KeyModifier.Shift, ftk.commandKeyModifier))
        self._addShortcut("Next", ftk.KeyShortcut(ftk.Key.PageDown, ftk.KeyModifier.Control))
        self._addShortcut("Prev", ftk.KeyShortcut(ftk.Key.PageUp, ftk.KeyModifier.Control))
        self._addShortcut("NextMediaReference", ftk.KeyShortcut(ftk.Key.M, ftk.KeyModifier.Shift))
        self._addShortcut("NextLayer", ftk.KeyShortcut(ftk.Key.Equals, ftk.KeyModifier.Control))
        self._addShortcut("PrevLayer", ftk.KeyShortcut(ftk.Key.Minus, ftk.KeyModifier.Control))
        self._addShortcut("Exit", ftk.KeyShortcut(ftk.Key.Q, ftk.commandKeyModifier))

        self._shortcutsUpdate(self._settingsModel.shortcuts)

        selfWeak = weakref.ref(self)
        self._filesObserver = djv.models.FilesModelItemListObserver(
            app.getFilesModel().observeFiles,
            lambda files: selfWeak()._filesUpdate(files))
        self._aObserver = djv.models.FilesModelItemObserver(
            app.getFilesModel().observeA,
            lambda item: selfWeak()._aUpdate(item))
        self._appWeak = appWeak
        self._playerObserver = tl.PlayerObserver(
            app.observePlayer(),
            lambda player: selfWeak()._playerUpdate(player))

    def _nextMediaReference(self):
        player = self._appWeak().observePlayer().get()
        if player is None:
            return
        # Cycle through the keys used by the timeline, starting from the
        # one in use. An unset key, which leaves the clips as they were
        # authored, is not part of the cycle.
        keys = player.mediaReferenceKeys
        if keys:
            try:
                next = (keys.index(player.mediaReferenceKey) + 1) % len(keys)
            except ValueError:
                next = 0
            player.mediaReferenceKey = keys[next]

    def _filesUpdate(self, files):
        enabled = len(files) > 0
        self.actions["SavePlaylist"].enabled = enabled
        self.actions["Close"].enabled = enabled
        self.actions["CloseAll"].enabled = enabled
        self.actions["Reload"].enabled = enabled
        # There is nothing to save, and nothing to close, until a file
        # is open. Opening a review stays available.
        self.actions["SaveReview"].enabled = enabled
        self.actions["SaveReviewAs"].enabled = enabled
        self.actions["CloseReview"].enabled = enabled
        self.actions["Next"].enabled = len(files) > 1
        self.actions["Prev"].enabled = len(files) > 1

    def _aUpdate(self, item):
        enabled = len(item.videoLayers) > 1 if item else False
        self.actions["NextLayer"].enabled = enabled
        self.actions["PrevLayer"].enabled = enabled

    def _playerUpdate(self, player):
        # There is nothing to cycle through unless the timeline uses more
        # than one media reference key.
        self.actions["NextMediaReference"].enabled = \
            len(player.mediaReferenceKeys) > 1 if player else False
